### ISSUE TITLE
Compatibility for packages with "irregular" version numbers

### ISSUE THREAD

**Issue by adrianliaw** [kennethreitz/pipenv#203]
Currently, there is a restriction that the version numbers of installed libraries have to be matching `^[\d]+\.[\d]+.*` (starting with this regex)

I'm trying to install pdfminer.six in my project, but due to the "irregular" version number that looks something like `20160614`, pipenv seemed to be failing to catch the version number and failed to write it properly into the lock file.
And it ended up looking like this:
```
"pdfminer.six": {
    "hash": "sha256:8897b7ca5e94329aef5b862e501834ea2c0460c9940c904fdbd07dba297c2553",
    "version": "=="
},
```
Which causes an installation failure while running `pipenv install`, that shows `Could not find a version that satisfies the requirement pdfminer.six==`

A way to fix this is to simply use another way to determine the version number when installing a package. That is, instead of looping through every component inside the downloaded distribution's file name separated by "-" and detect the version number using regex, we simply exclude the package name itself and file extension from the download fname, the leaving middle part is just the version number.

Will this approach be more suitable? Or is there a special reason for using the original approach?

**Comment by nateprewitt** [comment]
Thanks for raising this issue @adrianliaw. I agree the regex isn't tolerant of non-standard version numbers, but has this far been the most universal solution. We originally had a solution similar to your suggestion, but packages aren't guaranteed to start with their own name and they often use - and _ interchangeably in projects, even from one version to the next.

I do think this is an area that could use improvement but we need a solution that is more permissive of different version number patterns without mangling the majority of cases pipenv covers now.

**Comment by adrianliaw** [comment]
@nateprewitt Thanks for the explanation! But it seems like the file names of downloaded distributions such as tarballs and wheels, always start with the form of PKGNAME-VERSION, even though the module name you're going to use to import may not exactly be PKGNAME, but PKGNAME is exactly what you type in when you're trying to install. And according to [PEP426](https://www.python.org/dev/peps/pep-0426/#name), the package name MUST be case insensitive and MUST have no distinguishing between hyphens and underscores, so it wouldn't affect that much.

In addition, I also think that it might be better to use the official registered name of packages inside Pipfile, instead of using what you've typed in when you use `pipenv install PKGNAME`. e.g. If you used `pipenv install jinja2` to install Jinja2, pipenv should write `Jinja2` into Pipfile instead of `jinja2`. Running `pipenv install youtube-dl` should write `youtube_dl` instead of `youtube-dl` inside Pipfile.

One way to solve this problem is to directly make requests to PyPI and check the name of it. Or maybe using internal functionalities of pip, if that kind of functionalities exists?

**Comment by nateprewitt** [comment]
>One way to solve this problem is to directly make requests to PyPI and check the name of it. Or maybe using internal functionalities of pip, if that kind of functionalities exists? 

So we do already auto-normalize names from PyPI in `proper_name`. The problem is that depending on which pypi endpoint you hit and which version of the package you choose, you can end up with different results. We had an original implementation in 3.1.x that was based off of PEP426 and there were multiple issues opened because packages weren't returning consistent output. This gets particularly more complicated when you start mixing downloads from git which aren't required to adhere to the PEP426 filename specs.

Like I said above,
>I do think this is an area that could use improvement but we need a solution that is more permissive of different version number patterns without mangling the majority of cases pipenv covers now.

If you would like to write a patch that fixes this, I'll gladly review it. It will need to provide tests showing the issues listed in #90 (and the other 5 or 6 issue floating around) are still functioning though. I'm currently in the process of refactoring a few things that this will likely fall into but I'm not sure what the timeline is for a fix.

**Comment by nateprewitt** [comment]
This should be resolved with #218.
